### PR TITLE
UPDATE:  MAPIEv1 Backend -  Split .fit into .fit_single_estimator and fit_multi_estimator

### PR DESCRIPTION
**Pull request by jawadhussein462**
…_multi_estimators

# Description

This PR refactors the `MAPIE` ibrary by splitting the `.fit` method in `EnsembleRegressor` into two distinct methods: `.fit_single_estimator` and `.fit_multi_estimators`. The goal is to enhance modularity, readability, and maintainability of the codebase.

This refactoring is part of an ongoing effort to align the `MAPIE` library with modern software engineering practices. It ensures better scalability, maintainability, and adaptability for future features or modifications. The changes also address the growing need to handle single and multi-estimator workflows distinctly in the library.

1. **Refactoring `.fit` Logic**:
   - The `.fit` method is now divided into:
     - **`.fit_single_estimator`**: Handles the fitting logic for a single estimator.
     - **`.fit_multi_estimators`**: Handles the fitting logic for multiple estimators in an ensemble.

2. **Reorganization of `v0` and `v1` Components**:
   - **`v0` **:
     - `v0.conformlize`:
       Combines logic for `fit_multi_estimators` and calibration-based prediction (`predict_calib`).
     - `v0.fit`:
       Combines `fit_single_estimator` and `conformlize` to handle the full fitting process.
   - **`v1`**:
     - `v1.fit`:
       Only includes `fit_single_estimator`, streamlining the single-estimator fitting process.
     - `v1.conformlize`:
       Retains the structure of `v0.conformlize`, combining multi-estimator fitting and prediction calibration.

## Type of change

- New feature Non-breaking change which adds functionality

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/scikit-learn-contrib/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/scikit-learn-contrib/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/scikit-learn-contrib/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully and without warnings : `make doc`

### CODE REVIEW

File: mapie/estimator/regressor.py [hunk 1]
```diff
@@ -411,84 +411,41 @@ def predict_calib(
...
-        estimator = self.estimator
         n_samples = _num_samples(y)
+        estimators: List[RegressorMixin] = []
+
+        if self.cv == "prefit":
 
-        # Computation
-        if cv == "prefit":
-            single_estimator_ = estimator
+            # self.k_ is defined but not used
```
Comment (Valentin-Laurent) [inline mapie/estimator/regressor.py:428]: Obscure comment: to clarify ;)

File: mapie/estimator/regressor.py [hunk 2]
```diff
@@ -411,84 +411,41 @@ def predict_calib(
...
-
-        groups: Optional[ArrayLike] of shape (n_samples,)
-            Group labels for the samples used while splitting the dataset into
-            train/test set.
-
-            By default ``None``.
-
-        **fit_params : dict
-            Additional fit parameters.
+    ) -> RegressorMixin:
```
Comment (Valentin-Laurent) [inline mapie/estimator/regressor.py:421]:
```suggestion
    ) -> EnsembleRegressor:
```

File: mapie/estimator/regressor.py [hunk 3]
```diff
@@ -497,13 +454,44 @@ def fit(
...
-                if self.use_split_method_:
-                    single_estimator_ = estimators_[0]
 
-        self.single_estimator_ = single_estimator_
-        self.estimators_ = estimators_
+        self.estimators_ = estimators
+
+        return self
+
+    def fit(
```
Comment (Valentin-Laurent) [inline mapie/estimator/regressor.py:462]: I prefered `fit_single_estimator`: the purpose of the function is clearer

File: mapie/regression/regression.py [hunk 4]
```diff
@@ -513,12 +513,30 @@ def fit(
...
             The model itself.
         """
-        fit_params = kwargs.pop('fit_params', {})
-        predict_params = kwargs.pop('predict_params', {})
-        if len(predict_params) > 0:
-            self._predict_params = True
-        else:
-            self._predict_params = False
+
+        # Initialize the estimator
```
Comment (Valentin-Laurent) [inline mapie/regression/regression.py:517]: Remove comments from that section (if you think they are needed, it means that the functions name are not clear enough)

File: mapie/regression/regression.py [hunk 5]
```diff
@@ -513,12 +513,30 @@ def fit(
...
+        X, y, sample_weight, groups = self.init_fit(
+            X, y, sample_weight, groups, **kwargs
+        )
+
+        # Fit the prediction function
+        self.fit_single_estimator(X, y, sample_weight, groups)
+
+        # Conformlize the model:
+        self.conformlize(X, y, sample_weight, groups, **kwargs)
```
Comment (Valentin-Laurent) [inline mapie/regression/regression.py:526]: typo in conformalize

File: mapie_v1/regression.py [hunk 6]
```diff
@@ -320,7 +320,21 @@ def __init__(
...
+            estimator=self.estimator,
+            method=method,
+            cv=cv,
+            n_jobs=n_jobs,
+            verbose=verbose,
+            conformity_score=self.conformity_score,
+            random_state=random_state,
+        )
+
+        if isinstance(confidence_level, float):
```
Comment (Valentin-Laurent) [inline mapie_v1/regression.py:334]: Duplicated code

**Review by Valentin-Laurent — CHANGES_REQUESTED** [review]
Good job 💪🏻
Few comments

File: mapie/estimator/regressor.py (hunk 3 above)
Comment (jawadhussein462) [reply mapie/estimator/regressor.py]: Done!

File: mapie/regression/regression.py (hunk 5 above)
Comment (jawadhussein462) [reply mapie/regression/regression.py]: done

**Pull request merged**